File: src/wasm-ops/cumsum.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

#include "common.h"
#include "sum.h"
#include "utils/shape_utils.h"

// Wasm interop method
void cumsum_f32(void *data) {
  uint32_t *dataIndex = static_cast<uint32_t *>(data);
  uint32_t const argc = dataIndex[0];
  const float *x = PARAM_FLOAT_PTR(data, dataIndex[1]);
  const int32_t *dims = PARAM_INT32_PTR(data, dataIndex[2]);
  const int32_t rank = PARAM_INT32(data, dataIndex[3]);
  const int32_t axis = PARAM_INT32(data, dataIndex[4]);
  const bool exclusive = PARAM_BOOL(data, dataIndex[5]);
  const bool reverse = PARAM_BOOL(data, dataIndex[6]);

  float *output = PARAM_FLOAT_PTR(data, dataIndex[7]);
  cumsum_f32_imp(x, dims, rank, axis, exclusive, reverse, output);
}

// Core operator implementation
void cumsum_f32_imp(const float *X, const int32_t *dims, const int32_t rank,
                    int32_t axis, const bool exclusive, const bool reverse,
                    float *Y) {
  if (axis < 0) {
    axis = rank + axis;
  }

  // const index : number[] = new Array(y.dims.length).fill(0);
  size_t i = 0;
  std::vector<int32_t> dimsVector(dims, dims + rank);
  std::vector<int32_t> strides = ShapeUtils::compute_strides(dimsVector);
  size_t size = ShapeUtils::size_from_dims(dimsVector);

  if (reverse) {
    i = size - 1;
  }

  while (i < size && i >= 0) {

    size_t indexAtAxis = ShapeUtils::offset_to_index(strides, i, axis);

    size_t prevIndex = i + (reverse ? strides.at(axis) : -strides.at(axis));

    bool start = (indexAtAxis == 0 && !reverse) ||
                 (indexAtAxis == dimsVector.at(axis) && reverse);

    if (start && !exclusive) {
      Y[i] = X[i];
    } else if (start && exclusive) {
      Y[i] = 0;
    } else if (!start && !exclusive) {
      Y[i] = Y[prevIndex] + X[i];
    } else {
      Y[i] = Y[prevIndex] + X[prevIndex];
    }

    if (reverse) {
      i--;
    } else {
      i++;
    }
  }
}

File: lib/backends/cpu/ops/cumsum.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

import { Tensor } from '../../../tensor';
import { ShapeUtil } from '../../../util';
import { CpuInferenceHandler } from '../inference-handler';
import { CumSum } from '../../../ops/cumsum';

export class CpuCumSum extends CumSum {
  run(inferenceHandler: CpuInferenceHandler, inputs: Tensor[]): Tensor[] {
    const ax = inputs[1].integerData[0];
    const output = cumsum(inputs[0], ax, this.exclusive, this.reverse);
    return [output];
  }
}

export function cumsum(x: Tensor, axis: number, exclusive: boolean, reverse: boolean) {
  const y = new Tensor(x.dims, x.type);

  if (axis < 0) {
    axis = y.dims.length + axis;
  }

  const index: number[] = new Array(y.dims.length).fill(0);
  let i = 0;

  if (reverse) {
    i = y.data.length - 1;
    for (let j = 0; j < y.dims.length; j++) {
      index[j] = y.dims[j] - 1;
    }
  }

  while (i < y.data.length && i >= 0) {
    const prevIndex = updateIndex(index, axis, index[axis] + (reverse ? 1 : -1));

    const start = (index[axis] === 0 && !reverse) || (index[axis] === (y.dims[axis] - 1) && reverse);

    if (start && !exclusive) {
      y.set(index, x.get(index));
    } else if (start && exclusive) {
      y.set(index, 0);
    } else if (!start && !exclusive) {
      const prevValue = y.get(prevIndex) as number;
      y.set(index, prevValue + (x.get(index) as number));
    } else {
      const prevValue = y.get(prevIndex) as number;
      y.set(index, prevValue + (x.get(prevIndex) as number));
    }

    if (reverse) {
      ShapeUtil.decrementIndex(index, x.dims);
      i--;
    } else {
      ShapeUtil.incrementIndex(index, x.dims);
      i++;
    }
  }

  return y;
}

function updateIndex(index: number[], axis: number, value: number) {
  const result = index.slice();
  result[axis] = value;
  return result;
}
